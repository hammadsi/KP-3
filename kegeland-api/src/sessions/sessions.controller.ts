import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  UseGuards,
  UseInterceptors,
  ClassSerializerInterceptor,
} from '@nestjs/common';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';

import { FirebaseAuthGuard } from '../firebase/firebase-auth.guard';
import { Role } from '../roles/enums/role.enum';
import { Roles } from '../roles/roles.decorator';
import { RolesGuard } from '../roles/roles.guard';

import { SessionsService } from './sessions.service';
import { CreateSessionDto } from './dto/create-session.dto';
import { UpdateSessionDto } from './dto/update-session.dto';
import { ListSessionsDto } from './dto/list-sessions.dto';

@ApiTags('Sessions')
@Controller('sessions')
@UseInterceptors(ClassSerializerInterceptor)
@ApiBearerAuth('access-token')
@UseGuards(FirebaseAuthGuard, RolesGuard)
export class SessionsController {
  constructor(private readonly sessionsService: SessionsService) {}

  /**
   * Endpoint for fetching specific session by its ID
   * @param id of session
   * @returns Session object
   */
  @Get(':id')
  @Roles(Role.PATIENT, Role.PHYSICIAN)
  async findOne(@Param('id') id: string) {
    return this.sessionsService.findOne(id);
  }

  /**
   * Endpoint for fetching all sessions based on query parameters
   * @param filters query parameters passed in URL of request
   * @returns list of Session objects according to query parameters
   */
  @Get()
  @Roles(Role.PHYSICIAN)
  async findAll(@Query() filters: ListSessionsDto) {
    return this.sessionsService.findAll(filters);
  }

  /**
   * Endpoint for updating sessions object
   * @param id of session
   * @param updateSessionDto fields and data to be updated passed as body of request
   * @returns updated session object
   */
  @Put(':id')
  @Roles(Role.PATIENT)
  async update(
    @Param('id') id: string,
    @Body() updateSessionDto: UpdateSessionDto,
  ) {
    return this.sessionsService.update(id, updateSessionDto);
  }

  /**
   * Endpoint for creating session
   * @param createSessionDto data for session passed as body of request
   * @returns created session object
   */
  @Post()
  @Roles(Role.PATIENT)
  async create(@Body() createSessionDto: CreateSessionDto) {
    return this.sessionsService.create(createSessionDto);
  }

  /**
   * Endpoint for deleting session object
   * @param id id of session to be deleted
   * @returns Promise<void>
   */
  @Delete(':id')
  @Roles(Role.PATIENT, Role.PHYSICIAN)
  async delete(@Param('id') id: string) {
    return this.sessionsService.delete(id);
  }
}
