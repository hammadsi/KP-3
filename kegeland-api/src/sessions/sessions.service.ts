import { Injectable } from '@nestjs/common';
import { map } from 'lodash';
import { FirebaseService } from 'src/firebase/firebase.service';

import { timestamp } from '../utils/timestamp';

import { CreateSessionDto } from './dto/create-session.dto';
import { ListSessionsDto, SessionListItem } from './dto/list-sessions.dto';
import { UpdateSessionDto } from './dto/update-session.dto';
import { Session } from './entities/session.entity';

@Injectable()
export class SessionsService {
  constructor(private readonly firebaseService: FirebaseService) {}

  /**
   * Function for finding Session object by its ID
   * @param id of session to be found
   * @returns Session object
   */
  async findOneByPatientAndExercise(patientID: string, exerciseID: string): Promise<Session> {
    const snapshot = await this.firebaseService.firestore
      .collection('patients')
      .doc(patientID)
      .collection('gameSessions')
      .doc(exerciseID)
      .get();
    return {
      id: snapshot.id,
      ...snapshot.data(),
    } as Session;
  }

  /**
   * Function for finding all sessions based on query parameters
   * @param filters query parameters: name of sensor and/or userId
   * @returns list of all Sessions according to query parameters
   */
  async findAll(filters: ListSessionsDto): Promise<SessionListItem[]> {
    let query: any = this.firebaseService.firestore.collection('sessions');
    if ('sensor' in filters) {
      query = query.where('sensor', '==', filters.sensor);
    }
    if ('userId' in filters) {
      query = query.where('userId', '==', filters.userId);
    }
    const snapshots = await query.orderBy('createdAt', 'desc').get();
    return map(snapshots.docs, (doc) => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { data, ...spread } = doc.data();
      return {
        id: doc.id,
        ...spread,
      };
    });
  }

  /**
   * Function for creating new session in database
   * @param data fields and data for new sessions entry
   * @returns created Sessions object
   */
  async create(data: CreateSessionDto): Promise<Session> {
    const ts = timestamp();
    const docRef = await this.firebaseService.firestore
      .collection('sessions')
      .add({ ...data, createdAt: ts });
    return { id: docRef.id, ...data, createdAt: ts };
  }

  /**
   * Function for updating existing sessions object
   * @param id of Sessions to be updated
   * @param data fields and data to be changed
   * @returns updated Sessions object
   */
  async update(id: string, data: UpdateSessionDto): Promise<Partial<Session>> {
    await this.firebaseService.firestore
      .collection('sessions')
      .doc(id)
      .set(data, { merge: true });
    return { id, ...data };
  }

  /**
   * Function for deleting Sessions object
   * @param id of object to be deleted
   */
  async delete(id: string): Promise<void> {
    await this.firebaseService.firestore
      .collection('sessions')
      .doc(id)
      .delete();
  }
}
