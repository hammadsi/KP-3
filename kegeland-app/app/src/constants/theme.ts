import {cloneDeep, merge} from 'lodash';
import {
  DarkTheme as NavigationDarkTheme,
  DefaultTheme as NavigationLightTheme,
} from '@react-navigation/native';
import {
  DarkTheme as PaperDarkTheme,
  DefaultTheme as PaperLightTheme,
  Theme,
} from 'react-native-paper';

export type ThemeOverride = Theme['colors'] & {
  muted: string;
  elevation: {
    elevation1: string;
  };
  status: {
    success: string;
    alert: string;
    error: string;
  };
};

/**
 * The base override for theme
 */
const baseOverride: Partial<Omit<Theme, 'colors'>> = {
  roundness: 15,
};

/**
 * Color overrides for both light- and dark mode
 */
const sharedColorOverride: Partial<ThemeOverride> = {
  primary: '#00b0ff',
  status: {
    success: '#64dd17',
    alert: '#ffd600',
    error: '#dd2c00',
  },
};

/**
 * Color overrides for light mode
 */
const lightColorOverride: Partial<Theme['colors']> = merge(
  cloneDeep(sharedColorOverride),
  {
    muted: 'rgba(0, 0, 0, 0.54)',
    elevation: {
      elevation1: '#fff',
    },
  } as ThemeOverride,
);

/**
 * Color overrides for dark mode
 */
const darkColorOverride: Partial<Theme['colors']> = merge(
  cloneDeep(sharedColorOverride),
  {
    muted: 'rgba(229, 229, 231, 0.54)',
    elevation: {
      elevation1: '#272727',
    },
  } as ThemeOverride,
);

/**
 * The app's light theme
 */
export const lightTheme: any = merge(
  merge(NavigationLightTheme, PaperLightTheme),
  {
    ...baseOverride,
    colors: lightColorOverride,
  },
);

/**
 * The app's dark theme
 */
export const darkTheme: any = merge(
  merge(PaperDarkTheme, NavigationDarkTheme),
  {
    ...baseOverride,
    colors: darkColorOverride,
  },
);
