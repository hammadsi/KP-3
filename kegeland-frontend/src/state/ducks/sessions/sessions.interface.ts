import { SensorType } from '../sensors/sensors.interface';
import { GameSession } from '../wheelchairPatients/wheelchairPatients.interface';

export type SessionDataPoint = number | null;

export type LeanSession = {
  id: string;
  userId: string;
  sensor: SensorType;
  createdAt: number;
};

export type ViewSession = LeanSession | GameSession;

export interface Session extends LeanSession {
  data: Record<string, SessionDataPoint[]>;
}

export type SessionsState = {
  loading: boolean;
  session: Session | undefined;
  data: Session[];
  error: string | undefined;
};

export type FetchSessionsDto = {
  sensor?: SensorType;
  userId?: string;
};
