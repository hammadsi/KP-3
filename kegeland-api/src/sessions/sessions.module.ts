import { Module } from '@nestjs/common';

import { SessionsService } from './sessions.service';
import { SessionsController } from './sessions.controller';

@Module({
  imports: [],
  controllers: [SessionsController],
  providers: [SessionsService],
})
export class SessionsModule {}
