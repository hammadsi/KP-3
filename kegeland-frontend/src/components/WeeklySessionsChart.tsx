import { ChartData, ChartOptions } from 'chart.js';
import { Bar } from 'react-chartjs-2';
import 'chart.js/auto';
import { useTheme } from '@chakra-ui/react';
import { merge } from 'lodash';

import { ViewSession } from '../state/ducks/sessions/sessions.interface';
import { groupByWeek } from '../utils/session.utils';
import { CustomTheme } from '../types';

type WeeklySessionsChartProps = {
  sessions: ViewSession[];
  numWeeks: number;
};

const initChartData = (
  sessions: ViewSession[],
  numWeeks: number,
  color: string,
): ChartData<'bar'> => {
  const groups = groupByWeek(sessions, numWeeks);
  return {
    labels: Object.keys(groups),
    datasets: [
      {
        label: 'Exercises',
        data: Object.values(groups),
        backgroundColor: color,
      },
    ],
  };
};

const WeeklySessionsChart: React.FC<WeeklySessionsChartProps> = ({
  sessions,
  numWeeks,
}) => {
  const theme = useTheme<CustomTheme>();
  const chartData = initChartData(
    sessions,
    numWeeks,
    theme.colors.primary[600],
  );
  return (
    <div style={{ width: '99%' }}>
      <Bar
        height={200}
        options={merge(chartOptions, {
          plugins: {
            title: {
              display: true,
              text: `Exercises last ${numWeeks} weeks`,
            },
          },
        })}
        data={chartData}
      />
    </div>
  );
};

const chartOptions: ChartOptions<'bar'> = {
  responsive: true,
  plugins: {
    legend: {
      position: 'top' as const,
    },
  },
  scales: {
    y: {
      ticks: {
        precision: 0,
      },
    },
  },
};

export default WeeklySessionsChart;
