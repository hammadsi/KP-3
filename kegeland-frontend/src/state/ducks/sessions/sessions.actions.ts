import { createAsyncThunk } from '@reduxjs/toolkit';

import { apiCaller } from '../../../utils/apiCaller';

import { FetchSessionsDto, Session } from './sessions.interface';

export const fetchSessionById = createAsyncThunk(
  'sessions/fetchSessionById',
  async (id: string) =>
    apiCaller<Session>({
      url: `sessions/${id}`,
      method: 'GET',
    }),
);

export const fetchSessions = createAsyncThunk(
  'sessions/fetchSessions',
  async (params: FetchSessionsDto) =>
    apiCaller<Session[]>({
      url: 'sessions',
      method: 'GET',
      params,
    }),
);
