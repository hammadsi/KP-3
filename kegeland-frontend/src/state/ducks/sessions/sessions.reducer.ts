import { createSlice } from '@reduxjs/toolkit';
import { chain } from 'lodash';

import {
  isFulfilledAction,
  isPendingAction,
  isRejectedAction,
} from '../../../utils/thunk.utils';

import { fetchSessionById, fetchSessions } from './sessions.actions';
import { SessionsState } from './sessions.interface';

export const initialState: SessionsState = {
  loading: false,
  session: undefined,
  data: [],
  error: undefined,
};

const sessionsSlice = createSlice({
  name: 'sessions',
  initialState,
  reducers: {
    clearSessionsState: (state) => {
      state.loading = false;
      state.session = undefined;
      state.data = [];
      state.error = undefined;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchSessionById.fulfilled, (state, action) => {
        const { data, ...payload } = action.payload;
        state.session = {
          ...payload,
          data: chain(data).toPairs().sortBy(0).fromPairs().value(),
        };
      })
      .addCase(fetchSessions.fulfilled, (state, action) => {
        state.data = action.payload;
      })
      .addMatcher(
        (action) => isPendingAction(action, 'sessions'),
        (state) => {
          state.loading = true;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isFulfilledAction(action, 'sessions'),
        (state) => {
          state.loading = false;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isRejectedAction(action, 'sessions'),
        (state, { error }) => {
          state.loading = false;
          state.error = error.message;
        },
      );
  },
});

export const { clearSessionsState } = sessionsSlice.actions;

export default sessionsSlice.reducer;
