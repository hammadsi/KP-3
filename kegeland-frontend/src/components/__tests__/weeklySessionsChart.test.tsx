import { ChakraProvider } from '@chakra-ui/react';
import renderer from 'react-test-renderer';

import { sessionsMock } from '../mocks/sessions.mock';
import WeeklySessionsChart from '../WeeklySessionsChart';
import theme from '../../constants/theme.constants';

describe('Test weekly sessions chart', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ChakraProvider theme={theme}>
          <WeeklySessionsChart sessions={sessionsMock} numWeeks={123} />
        </ChakraProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
