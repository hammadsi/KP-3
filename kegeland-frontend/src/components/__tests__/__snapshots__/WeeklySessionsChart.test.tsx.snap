// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test weekly sessions chart renders correctly 1`] = `
Array [
  <div
    style={
      Object {
        "width": "99%",
      }
    }
  >
    <canvas
      height={200}
      role="img"
      width={300}
    />
  </div>,
  <span />,
]
`;
